### PR TITLE
Create an explicit structured grid

### PR DESCRIPTION
**Pull request by rodrigomologni**
### Overview

This pull request introduces the Explicit Structured Grid, available in VTK 9:
https://blog.kitware.com/introducing-explicit-structured-grids-in-vtk-and-paraview/

### Details

- Add `ExplicitStructuredGrid` to `pointset.py`
- Add `UnstructuredGrid.cast_to_explicit_structured_grid` to `pointset.py`
- Add `load_explicit_structured` to `examples.py`
- Add `create-explicit-structured-grid.py`
- Update `Common.cell_points` in `common.py`

### Examples

[examples.zip](https://github.com/pyvista/pyvista/files/5904754/examples.zip)

#### 1. Load an explicit structured grid:

```python
import pyvista as pv

grid = pv.ExplicitStructuredGrid('unisim.vtu')
grid.plot(color='w', show_edges=True, show_bounds=True)
```

image

#### 2. Show the hidden cells:

```python
import pyvista as pv

grid = pv.ExplicitStructuredGrid('unisim.vtu')
grid.show_cells()
grid.plot(color='w', show_edges=True, show_bounds=True)
```

image

#### 3. Compute the faces connectivity flags array:

```python
import pyvista as pv

grid = pv.ExplicitStructuredGrid('unisim.vtu')
grid.compute_connectivity()
grid.plot(scalars='ConnectivityFlags', show_edges=True)
```

image

#### 4. Compute an array with the number of connected cell faces:

```python
import pyvista as pv

grid = pv.ExplicitStructuredGrid('unisim.vtu')
grid.compute_connections()
grid.plot(scalars='number_of_connections', show_edges=True)
```

image

#### 5. Get the neighbors of a cell:

```python
import pyvista as pv

ind = 1824
grid = pv.ExplicitStructuredGrid('unisim.vtu')

plotter = pv.Plotter()
plotter.add_axes()

cell = grid.extract_cells(ind)
plotter.add_mesh(cell, color='lightcoral', show_edges=True, render=False)

ind = grid.neighbors(ind, rel='geometric')
cells = grid.extract_cells(ind)
plotter.add_mesh(cells, color='white', show_edges=True, render=False)

cells = cells.merge(cell)
cells = cells.cell_centers()
ind = cells.cell_arrays['vtkOriginalCellIds']
coords = grid.cell_coords(ind)
plotter.add_point_labels(cells.points, coords, always_visible=True)

plotter.show()
```
image

### CODE REVIEW

**Comment by akaszynski** [comment]
This feature will have to be added post 0.28.  Also, the following figure highlights that we need to enable backface culling:
image

**Comment by rodrigomologni** [comment]
> ... the following figure highlights that we need to enable backface culling:

@akaszynski I think this problem comes from VTK. I written the following code to render an **unstructured grid**:

```python
 1   import vtk
 2
 3   reader = vtk.vtkXMLUnstructuredGridReader()
 4   reader.SetFileName('unisim.vtu')
 5   reader.Update()
 6
 7   grid = reader.GetOutput()
 8   grid.GetCellData().RemoveArray('vtkGhostType')
 9
10   mapper = vtk.vtkDataSetMapper()
11   mapper.SetInputData(grid)
12   mapper.ScalarVisibilityOff()
13
14   actor = vtk.vtkActor()
15   actor.SetMapper(mapper)
16   actor.GetProperty().EdgeVisibilityOn()
17   actor.GetProperty().LightingOff()
18   actor.GetProperty().SetColor(1, 1, 1)
19
20   renderer = vtk.vtkRenderer()
21   renderer.AddActor(actor)
22   renderer.SetBackground(.3, .3, .3)
23
24   window = vtk.vtkRenderWindow()
25   window.AddRenderer(renderer)
26   window.SetSize(800, 600)
27 
28   interactor = vtk.vtkRenderWindowInteractor()
29   interactor.SetRenderWindow(window)
30   interactor.SetInteractorStyle(vtk.vtkInteractorStyleRubberBandPick())
31   interactor.Start()
```

image

After, I changed the code to render an **explicit structured grid** from an unstructured grid:

```python
 8   grid.GetCellData().RemoveArray('vtkGhostType')
 9
10   converter = vtk.vtkUnstructuredGridToExplicitStructuredGrid()
11   converter.SetInputData(grid)
12   converter.SetInputArrayToProcess(0, 0, 0, 1, 'BLOCK_I')
13   converter.SetInputArrayToProcess(1, 0, 0, 1, 'BLOCK_J')
14   converter.SetInputArrayToProcess(2, 0, 0, 1, 'BLOCK_K')
15   converter.GlobalWarningDisplayOff()
16   converter.Update()
17
18   grid = converter.GetOutput()
19
20   mapper = vtk.vtkDataSetMapper()
```

image

Download the code: [test.zip](https://github.com/pyvista/pyvista/files/5953565/test.zip)

**Comment by akaszynski** [comment]
I think part of the issue is the `extract_surface` filter doesn't actually extract the surface of the grid but instead extracts all the faces, including the interior faces.  This may have to be fixed on VTK's side.

**Comment by rodrigomologni** [comment]
I just remembered that I reported this error to the VTK developers almost a year ago. I think I was not clear in my explanation. I updated the conversation with this example. Let's wait for an answer. See: https://gitlab.kitware.com/vtk/vtk/-/issues/17783

**Comment by rodrigomologni** [comment]
Hi @akaszynski! Would you like me to keep [rodrigomologni:feat/esgrid](https://github.com/rodrigomologni/pyvista/tree/feat/esgrid) up to date with [pyvista:master](https://github.com/pyvista/pyvista)? If yes, how should I do this? Could I merge (squash) `master` into `feat/esgrid` and resolve the conflicts?

**Comment by akaszynski** [comment]
> Hi @akaszynski! Would you like me to keep [rodrigomologni:feat/esgrid](https://github.com/rodrigomologni/pyvista/tree/feat/esgrid) up to date with [pyvista:master](https://github.com/pyvista/pyvista)? If yes, how should I do this? Could I merge (squash) `master` into `feat/esgrid` and resolve the conflicts?

Merged for you since there's a bit of work dealing with the recent Common refactor as well as the `_vtk` refactor.  Noted a few problems:
- Doctests will fail on VTK8.  We need to find a way to skip these for our documentation build since it's done on VTK8.
- Unit tests aren't covering the entire class.  Many of the examples in the doc-strings can just be copied over to the unit tests.
- Some of the doctests had `np.int64`.  Was this intended?  I ended up removing it to get the tests to pass, but is it different on your system.

**Comment by rodrigomologni** [comment]
@akaszynski Thank you for merging the branches, resolving the conflicts and identifying the test problems. I will try to solve these problems along this week. About the use of `np.int64`, you're right, I used it to work on my system. Thanks for removing.

**Comment by rodrigomologni** [comment]
Hi @akaszynski!

> Doctests will fail on VTK8. We need to find a way to skip these for our documentation build since it's done on VTK8.

Unfortunately, I didn't find a way to skip doctest for VTK 8. I had to use `# doctest: +SKIP`.

> Unit tests aren't covering the entire class. Many of the examples in the doc-strings can just be copied over to the unit tests.

I created unit tests for all the `ExplicitStructuredGrid` methods. Rename or merge the tests as you wish, please.

File: examples/00-load/create-explicit-structured-grid.py
Comment (akaszynski) [inline examples/00-load/create-explicit-structured-grid.py:7]:
```suggestion
Create an explicit structured grid from NumPy arrays.
```

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:830]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid.hide_cells(range(80, 120))  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid = grid.cast_to_unstructured_grid()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid = grid.cast_to_explicit_structured_grid()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP
```

Our sphinx docs will split these up into separate examples and it will look nicer when it renders.

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1224]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid.hide_cells(range(80, 120))  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid = grid.cast_to_unstructured_grid()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid = grid.cast_to_explicit_structured_grid()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP
```

See comment above.

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1266]:
```suggestion
        >>> import pyvista as pv
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.hide_cells(range(80, 120))  # doctest: +SKIP
        >>> grid.save('grid.vtu')  # doctest: +SKIP

        >>> grid = pv.ExplicitStructuredGrid('grid.vtu')  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid.show_cells()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP
```
See comment above.

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1294]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
```

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1337]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.hide_cells(range(80, 120))  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP

        >>> grid.show_cells()  # doctest: +SKIP
        >>> grid.plot(color='w', show_edges=True, show_bounds=True)  # doctest: +SKIP
```

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1375]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.dimensions  # doctest: +SKIP
```

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1403]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.hide_cells(range(80, 120))  # doctest: +SKIP
        >>> grid.bounds  # doctest: +SKIP
        [0.0, 80.0, 0.0, 50.0, 0.0, 6.0]

        >>> grid.visible_bounds  # doctest: +SKIP
        [0.0, 80.0, 0.0, 50.0, 0.0, 4.0]
```

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1444]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.cell_id((3, 4, 0))  # doctest: +SKIP
        19

        >>> coords = [(3, 4, 0),
        ...           (3, 2, 1),
        ...           (1, 0, 2),
        ...           (2, 3, 2)]
        >>> grid.cell_id(coords)  # doctest: +SKIP
        array([19, 31, 41, 54])
```

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1492]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.cell_coords(19)  # doctest: +SKIP
        (3, 4, 0)

        >>> grid.cell_coords((19, 31, 41, 54))  # doctest: +SKIP
        array([[3, 4, 0],
               [3, 2, 1],
               [1, 0, 2],
               [2, 3, 2]])
```

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1531]:
```suggestion
        >>> import pyvista as pv
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
```

File: pyvista/core/pointset.py
Comment (rodrigomologni) [reply pyvista/core/pointset.py:830]: You're right! I will consider your suggestion for my next codes. Thanks!

File: pyvista/core/pointset.py
Comment (akaszynski) [inline pyvista/core/pointset.py:1694]:
```suggestion
        >>> from pyvista import examples
        >>> grid = examples.load_explicit_structured()  # doctest: +SKIP
        >>> grid.compute_connections()  # doctest: +SKIP
        >>> grid.plot(show_edges=True)  # doctest: +SKIP
```

**Comment by akaszynski** [comment]
You can add the suggestions to a batch to click less :)

**Comment by rodrigomologni** [comment]
> You can add the suggestions to a batch to click less :)

I wanted to, but is not enable for me! :disappointed_relieved:

image

**Comment by akaszynski** [comment]
You have to do it from the "files" tab.  This took me a long time to figure out as well.

**Comment by rodrigomologni** [comment]
> You have to do it from the "files" tab. This took me a long time to figure out as well.

Hummmm... Thanks for the info! I will use this feature from now on.

**Review by akaszynski — APPROVED** [review]
I made some final fixes to wrap the docstrings of many functions according to the [numpydoc Style Guide](https://numpydoc.readthedocs.io/en/latest/format.html).  There was also a few misses in coverage that were easy to fix (and one outside of the scope of this PR).

**Comment by akaszynski** [comment]
I'll leave it up to you to merge.

**Comment by rodrigomologni** [comment]
@akaszynski Perfect! Thanks for your contributions!

I thought that all code could exceed 79 characters, to a max of 90 to 100 characters. I will produce docstrings of up to 75 characters from now on, as recommended in **numpydoc Style Guide**. Thanks for the information!

**Comment by akaszynski** [comment]
> I thought that all code could exceed 79 characters, to a max of 90 to 100 characters

Quite right in regards to our [Contributing](https://docs.pyvista.org/extras/developer_notes.html) page.  I should add additional notes regarding the numpy style guide.

Thanks for your work!

**Pull request merged**
